common/common/src/main/java/io/helidon/build/common/FileUtils.java

```diff
@@ -747,6 +750,67 @@ public static Path zip(Path zip, Path directory) {
         }
     }
 
+    /**
+     * Unzip a zip file.
+     *
+     * @param zip       source file
+     * @param directory target directory
+     */
+    public static void unzip(Path zip, Path directory) {
+        try (FileSystem fs = newZipFileSystem(zip)) {
+            if (!Files.exists(directory)) {
+                Files.createDirectory(directory);
+            }
+            Path root = fs.getRootDirectories().iterator().next();
+            Files.walk(root)
+                    .filter(p -> !p.equals(root))
+                    .forEach(file -> {
+                        Path filePath = directory.resolve(Path.of(file.toString().substring(1)));
+                        try {
+                            if (Files.isDirectory(file)) {
+                                Files.createDirectories(filePath);
+                            } else {
+                                Files.copy(file, filePath);
+                            }
+                        } catch (IOException ioe) {
+                            throw new UncheckedIOException(ioe);
+                        }
+                    });
+        } catch (IOException | UncheckedIOException e) {
+            //if zipFileSystem fails, try to unzip with ZipStream
+            unzip0(zip, directory);
```

Comment:
Can you explain this fallback ? Why would that work with ZipInputStream if it failed here ?!

Reply:
I don't know the exact reason, but in practice, a zip file without END header cannot be unzip with `FileSystem` but can be with stream. One solution I was thinking about is to update the zip file containing `mvn` distribution. But I do not have access to this location.

Reply:
Can you give me an example of a file that cannot be unzipped with nio ?

Reply:
Ok the end of stream seems to be caused by the implementation of `TestUtils.downloadFileFromUrl`:

The code below does not guarantee that all bytes have been transferred.
```java
fileOutputStream.getChannel().transferFrom(readableByteChannel, 0, Long.MAX_VALUE);
```

Instead you can use this:
```java
InputStream is = connection.getInputStream();
OutputStream os = Files.newOutputStream(destination);
is.transferTo(os);
```